src/dotnet/ShellShim/ShellShimMaker.cs

```diff
@@ -23,10 +25,37 @@ public class ShellShimMaker : IShellShimMaker
         public ShellShimMaker(string pathToPlaceShim)
         {
             _pathToPlaceShim =
-                pathToPlaceShim ?? throw new ArgumentNullException(nameof(pathToPlaceShim));
+                pathToPlaceShim ??
+                throw new ArgumentNullException(nameof(pathToPlaceShim));
         }
 
         public void CreateShim(FilePath packageExecutable, string shellCommandName)
+        {
+            var resourceManager = new TransactionResourceManager(
+                commit: () => { PlaceShim(packageExecutable, shellCommandName); },
```

Comment:
So in my experimentation, a transaction which throws from `Commit` does not rollback any enlistment; instead the exception goes uncaught by the transaction manager.

I think the proper place to do the work is in `Prepare`. `Commit` should be treated as a notification that the transaction is committed; it should not fail, only do whatever internal bookmarking is needed when the transaction is committed.

Reply:
Also, think about what happens when a caller calls `CreateShim` twice with the same `shellCommandName` parameter.  We enlist two notifications that check for the existence of the same file; both checks will succeed, however, the second actual commit will fail because the files exist.

Either the work needs be done in `CreateShim` and it enlists only undo actions (my preference), or the work needs to be done in `Prepare`.

Reply:
Agree, and this step will not fail. All is checked out during prepare type. This step is just copy the embedded file and write config.

If this a step, the roll back could delete the conflicted file(what used to be there). To avoid it, we need write and rename. This is really no too much different than copy embedded file

Reply:
Commit is empty now